In [252]:
import nltk
import re
from nltk.corpus import *
from collections import defaultdict, Counter
import random

__☼ Recall Strunk and White's prohibition against sentence-initial however used to mean "although". Do a web search for however used at the start of the sentence. How widely used is this construction?__

In [7]:
however_sents = [sent for sent in brown.sents() if sent[0].lower() == "however"]
for sent in however_sents:
    print(*sent)

However , the jury said it believes `` these two offices should be combined to achieve greater efficiency and reduce the cost of administration '' .
however , the first belief stood for entire revision with a new third point added to the list .
However , this would require a lengthening of the season from thirteen to fourteen weeks .
However , Mr. Parichy and his bride will go to Vero Beach on their wedding trip , and will stay in the John G. Beadles' beach house .
However , the confession , which was the only evidence against him , was retracted before the trial .
However , three of the managers did say that they would agree to attend the proposed meeting if all of the other managers decided to attend .
However , compared with March 1960 new car sales of 4,441 , this March was off 23 per cent .
However , Peterson , president of Bell & Howell , warned 800 U.S. marketing leaders attending a national conference at the Ambassador , that the future will belong to the industrialist of creat

However , there is always the possibility that chance will make demands the dancers find impossible to execute .
However , for convenience we will stick to the idea that information can be classified according to Table 1 .
However , it is important to trace the philosophy of the French Revolution to its sources to understand the common democratic origin of individualism and socialism and the influence of the latter on the former .
However , the crisis occurs precisely as a weakening of spiritual forces .
However , I confess my hope that I will be innocent again , not with a pristine , accidental innocence , but rather with an innocence achieved by the slow cutting away of the flesh to reach the bone .
However , it was not of innocence in general that I was speaking , but of perhaps the frailest and surely the least important side of it which is innocence in romantic love .
However , in the context of legal and civic policy , these controversies are less than novel .
However , the Gener

However , no diagram of her brain patterns , no early I.Q. tests recorded certain essential facts about Helva that Central must eventually learn .
However , when there's a job to be done , I'm a monstrosity of grim determination , I like to think .
However while in his cell awaiting trial for theft , Barco , in a fit of apprehension , made an attempt to take his own life .
However , such a hotbed of gossip had grown up during the trial , that every precaution had to be taken to keep my visit from being whispered to the world , Society , and even , alas , to my own mother .
However one looks at it , therefore , I'd say that your horoscope for this autumn is the reverse of rosy .


__☼ Consider the sentence _Kim arrived or Dana left and everyone cheered._ Write down the parenthesized forms to show the relative scope of and and or. Generate tree structures corresponding to both of these interpretations.__

__☼ The Tree class implements a variety of other useful methods. See the Tree help documentation for more details, i.e. import the Tree class and then type help(Tree).__

In [10]:
help(nltk.Tree)

Help on class Tree in module nltk.tree:

class Tree(builtins.list)
 |  Tree(node, children=None)
 |  
 |  A Tree represents a hierarchical grouping of leaves and subtrees.
 |  For example, each constituent in a syntax tree is represented by a single Tree.
 |  
 |  A tree's children are encoded as a list of leaves and subtrees,
 |  where a leaf is a basic (non-tree) value; and a subtree is a
 |  nested Tree.
 |  
 |      >>> from nltk.tree import Tree
 |      >>> print(Tree(1, [2, Tree(3, [4]), 5]))
 |      (1 2 (3 4) 5)
 |      >>> vp = Tree('VP', [Tree('V', ['saw']),
 |      ...                  Tree('NP', ['him'])])
 |      >>> s = Tree('S', [Tree('NP', ['I']), vp])
 |      >>> print(s)
 |      (S (NP I) (VP (V saw) (NP him)))
 |      >>> print(s[1])
 |      (VP (V saw) (NP him))
 |      >>> print(s[1,1])
 |      (NP him)
 |      >>> t = Tree.fromstring("(S (NP I) (VP (V saw) (NP him)))")
 |      >>> s == t
 |      True
 |      >>> t[1][1].set_label('X')
 |      >>> t[1][1].label()
 

__☼ In this exercise you will manually construct some parse trees.__

Write code to produce two trees, one for each reading of the phrase _old men and women_

Encode any of the trees presented in this chapter as a labeled bracketing and use nltk.Tree() to check that it is well-formed. Now use draw() to display the tree.

As in (a) above, draw a tree for The woman saw a man last Thursday.

In [11]:
first = nltk.Tree.fromstring("(S (NP (JJ old) (N men)) (CC and) (NP (N women)))")
second = nltk.Tree.fromstring("(S (NP (JJ old) (NP (N men)) (CC and) (N women)))")

In [15]:
first.draw()
second.draw()

__☼ Write a recursive function to traverse a tree and return the depth of the tree, such that a tree with a single node would have depth zero. (Hint: the depth of a subtree is the maximum depth of its children, plus one.)__

In [17]:
print(treebank.parsed_sents()[0])

(S
  (NP-SBJ
    (NP (NNP Pierre) (NNP Vinken))
    (, ,)
    (ADJP (NP (CD 61) (NNS years)) (JJ old))
    (, ,))
  (VP
    (MD will)
    (VP
      (VB join)
      (NP (DT the) (NN board))
      (PP-CLR (IN as) (NP (DT a) (JJ nonexecutive) (NN director)))
      (NP-TMP (NNP Nov.) (CD 29))))
  (. .))


In [60]:
def traverse_tree(tree):
    depth = 0
    for subtree in tree:
        if len(subtree) == 1:
            depth += 1
        else:
            traverse_tree(subtree)
    return depth   

In [62]:
## WILL COME BACK LATER, I DID NOT GET THE QUESTION

__☼ Analyze the A.A. Milne sentence about Piglet, by underlining all of the sentences it contains then replacing these with S (e.g. the first sentence becomes S when:lx` S). Draw a tree structure for this "compressed" sentence. What are the main syntactic constructions used for building such a long sentence?__

_In after-years he liked to think_ that _he had been in Very Great Danger during the Terrible Flood, but the only danger 
he had really been in was the last half-hour of his imprisonment_, when _Owl, who had just flown up, sat on a branch of his 
tree to comfort him, and told him a very long story_ about _an aunt who had once laid a seagull’s egg by mistake_, and _the story went on and on, rather like this sentence, until Piglet who was listening out of his window without much hope, went to sleep quietly and naturally, slipping slowly out of the window towards the water until he was only hanging on by his toes_, at which moment, luckily, _a sudden loud squawk from Owl, which was really part of the story, being what his aunt said, woke the Piglet up and just gave him time to jerk himself back into safety and say, “How interesting, and did she?”_ when — well, _you can imagine his joy when at last he saw the good ship, Brain of Pooh (Captain, C. Robin; 1st Mate, P. Bear) coming over the sea to rescue him._

__☼ Can the grammar in grammar1 be used to describe sentences that are more than 20 words in length?__

In [73]:
grammar1 = nltk.CFG.fromstring("""
  S -> NP VP
  VP -> V NP | V NP PP
  PP -> P NP
  V -> "saw" | "ate" | "walked"
  NP -> "John" | "Mary" | "Bob" | Det N | Det N PP
  Det -> "a" | "an" | "the" | "my"
  N -> "man" | "dog" | "cat" | "telescope" | "park" | "mat" | "shop" | "pijamas"
  P -> "in" | "on" | "by" | "with" | "behind" | "in front of"
  """)

In [74]:
# It could be even longer by adding more PPs

rd_parser = nltk.RecursiveDescentParser(grammar1)
sent = "a man with a telescope saw the cat in the park on the mat behind the dog by the shop with my pijamas"

In [77]:
print(next(rd_parser.parse(sent.split())))

(S
  (NP (Det a) (N man) (PP (P with) (NP (Det a) (N telescope))))
  (VP
    (V saw)
    (NP
      (Det the)
      (N cat)
      (PP
        (P in)
        (NP
          (Det the)
          (N park)
          (PP
            (P on)
            (NP
              (Det the)
              (N mat)
              (PP
                (P behind)
                (NP
                  (Det the)
                  (N dog)
                  (PP
                    (P by)
                    (NP
                      (Det the)
                      (N shop)
                      (PP (P with) (NP (Det my) (N pijamas))))))))))))))


__◑ You can modify the grammar in the recursive descent parser demo by selecting Edit Grammar in the Edit menu. Change the second expansion production, namely NP -> Det N PP, to NP -> NP
PP. Using the Step button, try to build a parse tree. What happens?__

In [79]:
# It fails to parse correctly, infinitely expanding the NP to NP PP
nltk.app.rdparser()

__◑ Extend the grammar in grammar2 with productions that expand prepositions as intransitive, transitive and requiring a PP complement. Based on these productions, use the method of the preceding exercise to draw a tree for the sentence Lee ran away home.__

In [82]:
grammar2 = nltk.CFG.fromstring("""
  S  -> NP VP
  NP -> Det Nom | PropN
  Nom -> Adj Nom | N
  VP -> VI Adj | VT NP | VS S | VT NP PP
  PP -> P NP
  PropN -> 'Buster' | 'Chatterer' | 'Joe' | 'Lee'
  Det -> 'the' | 'a'
  N -> 'bear' | 'squirrel' | 'tree' | 'fish' | 'log' | 'dog' | 'cat'
  Adj  -> 'angry' | 'frightened' |  'little' | 'tall'
  VI -> 'was'
  VT -> 'chased'  | 'saw'
  VS -> 'said' | 'thought' | 'put'
  P -> 'on'
  """)

In [89]:
sent_2 = "the dog chased the cat"
ungram_sent = "the squirrel was Buster was angry"
parser = nltk.ChartParser(grammar2)
for parse in parser.parse(sent_2.split()):
    print(parse)

(S
  (NP (Det the) (Nom (N dog)))
  (VP (VT chased) (NP (Det the) (Nom (N cat)))))


In [90]:
# No parse generated
for parse in parser.parse(ungram_sent.split()):
    print(parse)

__◑ Pick some common verbs and complete the following tasks:__

Write a program to find those verbs in the Prepositional Phrase Attachment Corpus nltk.corpus.ppattach. Find any cases where the same verb exhibits two different attachments, but where the first noun, or second noun, or preposition, stay unchanged (as we saw in our discussion of syntactic ambiguity in 2)

Devise CFG grammar productions to cover some of these cases.

In [8]:
attachments = ppattach.attachments("training")

In [13]:
atts = defaultdict(lambda : defaultdict(set))
for at in attachments:
    atts[at.verb][at.attachment].add((at.noun1, at.prep, at.noun2))

In [14]:
# The list of verbs that can be both "V" and "N" attachment type, and where the noun1, noun2 and prep are the same
for verb, att_dict in atts.items():
    if len(att_dict["N"].intersection(att_dict["V"])) > 1:
        print(verb, (att_dict["N"].intersection(att_dict["V"])))

is {('reporter', 'in', 'bureau'), ('writer', 'in', 'York')}
have {('impact', 'on', 'results'), ('impact', 'on', 'market')}
had {('access', 'to', 'AZT'), ('effect', 'on', 'market'), ('loss', 'in', 'quarter'), ('change', 'in', 'earnings'), ('sales', 'of', 'million')}
keep {('pressure', 'on', 'government'), ('pace', 'with', 'inflation')}
make {('payments', 'on', 'debt'), ('payment', 'on', 'million')}
reported {('earnings', 'of', 'million'), ('loss', 'for', 'quarter'), ('earnings', 'for', 'companies')}
paid {('price', 'for', 'incentives'), ('attention', 'to', 'comments'), ('500,000', 'in', 'fines')}
posted {('drop', 'in', 'earnings'), ('gain', 'in', 'profit'), ('drop', 'in', 'profit')}
earned {('million', 'on', 'revenue'), ('million', 'on', 'sales')}


In [19]:
# sanity check
[ins for ins in attachments if ins.verb == "earned" and ins.noun1 == "million" and ins.prep == "on" and ins.noun2 == "revenue"]

[PPAttachment(sent='2927', verb='earned', noun1='million', prep='on', noun2='revenue', attachment='V'),
 PPAttachment(sent='3509', verb='earned', noun1='million', prep='on', noun2='revenue', attachment='V'),
 PPAttachment(sent='16034', verb='earned', noun1='million', prep='on', noun2='revenue', attachment='V'),
 PPAttachment(sent='16062', verb='earned', noun1='million', prep='on', noun2='revenue', attachment='V'),
 PPAttachment(sent='23700', verb='earned', noun1='million', prep='on', noun2='revenue', attachment='N'),
 PPAttachment(sent='24547', verb='earned', noun1='million', prep='on', noun2='revenue', attachment='V'),
 PPAttachment(sent='35187', verb='earned', noun1='million', prep='on', noun2='revenue', attachment='V')]

I would love to see the sentences from above examples of "earned million on revenue", but I could not find where to access the sentences

__◑ Write a program to compare the efficiency of a top-down chart parser compared with a recursive descent parser (4). Use the same grammar and input sentences for both. Compare their performance using the timeit module (see 4.7 for an example of how to do this).__

In [142]:
my_sent = "I thought Joe said Sam believes she saw a black cat on the mat in the park"

In [147]:
my_grammar = nltk.CFG.fromstring("""
  S  -> NP VP
  NP -> Det Nom | PropN
  Nom -> Adj Nom | N
  VP -> V Adj | V NP | V S | V NP PP | VP VP
  PP -> P NP
  PropN -> 'I' | 'Joe' | 'Sam' | 'she'
  Det -> 'the' | 'a'
  N -> 'Australia' | 'glasses' | 'cat' | 'mat' | 'park'
  Adj  -> 'black' | 'big'
  V -> 'said' | 'thought' | 'saw' | 'believes'
  P -> 'on' | 'in'
""")

In [148]:
top_down_chart = nltk.ChartParser(my_grammar)
recursive = nltk.RecursiveDescentParser(my_grammar)

In [149]:
%%timeit
top_down_chart.parse(my_sent.split())

10.5 ms ± 1.79 ms per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [150]:
%%timeit
recursive.parse(my_sent.split())

16.7 µs ± 7.74 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


__◑ Compare the performance of the top-down, bottom-up, and left-corner parsers using the same grammar and three grammatical test sentences. Use timeit to log the amount of time each parser takes on the same sentence. Write a function that runs all three parsers on all three sentences, and prints a 3-by-3 grid of times, as well as row and column totals. Discuss your findings.__

In [5]:
sent1 = "I thought Joe said Sam believes she saw a black cat on the mat in the park"
sent2 = "the cat chased the dog"
sent3 = "I bought an ice-cream in the park"

In [6]:
my_grammar_2 = nltk.CFG.fromstring(""" 
  S  -> NP VP | S CC S
  NP -> Det Nom | PropN
  Nom -> Adj Nom | N
  VP -> V Adj | V NP | V S | V NP PP | VP VP
  PP -> P NP
  PropN -> 'I' | 'Joe' | 'Sam' | 'she'
  Det -> 'the' | 'a' | 'an'
  N -> 'Australia' | 'glasses' | 'cat' | 'mat' | 'park' | 'ice-cream' | 'dog'
  Adj  -> 'black' | 'big'
  V -> 'said' | 'thought' | 'saw' | 'believes' | 'love' | 'bought' | 'chased'
  P -> 'on' | 'in'
  CC -> 'because' | 'and' | 'but'
""")

top_down_chart_2 = nltk.ChartParser(my_grammar_2)
recursive_2 = nltk.RecursiveDescentParser(my_grammar_2)

In [ ]:
import timeit

matrix_times = [[None for _ in range(3)] for _ in range(2)]
for idx, sent in enumerate([sent1, sent2, sent3]):
    for idx_p, parser in enumerate([top_down_chart_2, recursive_2]):
        matrix_times[idx_p][idx] = timeit.timeit(lambda : parser.parse(sent.split()))
print(matrix_times)

__◑ Read up on "garden path" sentences. How might the computational work of a parser relate to the difficulty humans have with processing these sentences? http://en.wikipedia.org/wiki/Garden_path_sentence__

In [ ]:
# TODO

__◑ To compare multiple trees in a single window, we can use the draw_trees() method. Define some trees and try it out:__

In [58]:
from nltk.draw.tree import draw_trees
draw_trees(treebank.parsed_sents()[0], treebank.parsed_sents()[1], treebank.parsed_sents()[2])  

__◑ Using tree positions, list the subjects of the first 100 sentences in the Penn treebank; to make the results easier to view, limit the extracted subjects to subtrees whose height is 2.__

In [75]:
for tree in treebank.parsed_sents()[:100]:
    for subtree in tree.subtrees():
        if subtree.label() == ("NP-SBJ") and subtree.height() == 3:
            print(subtree.flatten())

(NP-SBJ Mr. Vinken)
(NP-SBJ *-1)
(NP-SBJ *)
(NP-SBJ researchers)
(NP-SBJ it)
(NP-SBJ *T*-1)
(NP-SBJ researchers)
(NP-SBJ *T*-2)
(NP-SBJ the latest results)
(NP-SBJ *)
(NP-SBJ A Lorillard spokewoman)
(NP-SBJ This)
(NP-SBJ We)
(NP-SBJ anyone)
(NP-SBJ asbestos)
(NP-SBJ There)
(NP-SBJ *T*-3)
(NP-SBJ We)
(NP-SBJ users)
(NP-SBJ Dr. Talcott)
(NP-SBJ The Lorillard spokeswoman)
(NP-SBJ *)
(NP-SBJ the company)
(NP-SBJ *T*-4)
(NP-SBJ *)
(NP-SBJ the researchers)
(NP-SBJ The morbidity rate)
(NP-SBJ *T*-5)
(NP-SBJ Dr. Talcott)
(NP-SBJ *-1)
(NP-SBJ he)
(NP-SBJ *)
(NP-SBJ The finding)
(NP-SBJ *T*-6)
(NP-SBJ the U.S.)
(NP-SBJ Dr. Talcott)
(NP-SBJ The U.S.)
(NP-SBJ *T*-7)
(NP-SBJ Dr. Mossman)
(NP-SBJ the Environmental Protection Agency)
(NP-SBJ *T*-8)
(NP-SBJ Workers)
(NP-SBJ *)
(NP-SBJ Workers)
(NP-SBJ *T*-1)
(NP-SBJ exhaust fans)
(NP-SBJ There)
(NP-SBJ *-1)
(NP-SBJ these events)
(NP-SBJ It)
(NP-SBJ *-1)
(NP-SBJ portfolio managers)
(NP-SBJ Compound yields)
(NP-SBJ the current yield)
(NP-SBJ *-1)
(NP-SB

__◑ Inspect the Prepositional Phrase Attachment Corpus and try to suggest some factors that influence PP attachment.__

In [253]:
all_n_ins = [(ins.noun1, ins.prep) for ins in attachments if ins.attachment == "N"]
all_v_ins = [(ins.verb, ins.prep) for ins in attachments if ins.attachment == "V"]

In [254]:
print(*all_n_ins[:10])
print(*all_v_ins[:10])

('chairman', 'of') ('director', 'of') ('percentage', 'of') ('asbestos', 'in') ('team', 'of') ('paper', 'for') ('three', 'with') ('finding', 'among') ('one', 'of') ('standard', 'of')
('join', 'as') ('using', 'in') ('bring', 'to') ('dumped', 'into') ('mixed', 'in') ('retain', 'for') ('pour', 'into') ('raise', 'to') ('been', 'for') ('registered', 'in')


In [272]:
my_dict = defaultdict(lambda:list())
tuples = []
all_attachments_N = []
for ins in attachments:
    if (ins.noun1, ins.prep) in tuples:
        all_attachments_N.append(ins.attachment)
    if ins.attachment == "N":
        my_dict["N"].append((ins.noun1, ins.prep))
        tuples.append((ins.noun1, ins.prep))

In [273]:
Counter(all_attachments_N)

Counter({'N': 6397, 'V': 1991})

In [274]:
my_dict_2 = defaultdict(lambda:list())
tuples_2 = []
all_attachments_V = []
for ins in attachments:
    if (ins.verb, ins.prep) in tuples_2:
        all_attachments_V.append(ins.attachment)
    if ins.attachment == "V":
        my_dict_2["V"].append((ins.verb, ins.prep))
        tuples_2.append((ins.verb, ins.prep))

In [275]:
Counter(all_attachments_V)

Counter({'N': 2349, 'V': 5199})

So, I just tried to show that if we take a list of all N attachments and extract possible attachments for that list's (noun1 and prep), we get the proportion of N (6397) higher than V (1991). In other words, my guess is that N attachments heavily rely on noun1 and prep, while V attachments rely more of verb and prep.

In [278]:
# Let us try out the same thing with N attachments, but with same (noun2 and prep). 
# My guess is that the number of V and N attachments in Counter will be more equally distributed
# since (noun2 and prep) are not strong predictors of the N type attachment
my_dict_3 = defaultdict(lambda:list())
tuples_3 = []
all_attachments_N2 = []
for ins in attachments:
    if (ins.noun2, ins.prep) in tuples_3:
        all_attachments_N2.append(ins.attachment)
    if ins.attachment == "N":
        my_dict_3["N"].append((ins.noun2, ins.prep))
        tuples_3.append((ins.noun2, ins.prep))
Counter(all_attachments_N2)

Counter({'N': 4966, 'V': 2016})

__◑ In this section we claimed that there are linguistic regularities that cannot be described simply in terms of n-grams. Consider the following sentence, particularly the position of the phrase "in his turn". Does this illustrate a problem for an approach based on n-grams?__

    What was more, the in his turn somewhat youngish Nikolay Parfenovich also turned out to be the only person in the entire world to acquire a sincere liking to our "discriminated-against" public procurator. (Dostoevsky: The Brothers Karamazov)

In [ ]:
# TODO

__◑ Write a recursive function that produces a nested bracketing for a tree, leaving out the leaf nodes, and displaying the non-terminal labels after their subtrees. So the above example about Pierre Vinken would produce: [[[NNP NNP]NP , [ADJP [CD NNS]NP JJ]ADJP ,]NP-SBJ MD [VB [DT NN]NP [IN [DT JJ NN]NP]PP-CLR [NNP CD]NP-TMP]VP .]S Consecutive categories should be separated by space.__

In [249]:
def bracketing(tree, parents = []):
    
    parents.append(tree.label())
    
    l = len(tree) - 1
    for child in tree:
        if child.height() == 2:
            print(child.label(), end = "")
            if l > 0:
                print(" ", end = "")
                l -= 1
        elif isinstance(child, nltk.Tree):
            print("[", end = "")
            bracketing(child, parents)
    print("]" + parents.pop(), end = " ")

In [251]:
bracketing(treebank.parsed_sents()[0])

[[NNP NNP]NP , [[CD NNS]NP JJ ]ADJP , ]NP-SBJ [MD [VB [DT NN]NP [IN [DT JJ NN]NP ]PP-CLR [NNP CD]NP-TMP ]VP ]VP . ]S 

__◑ Download several electronic books from Project Gutenberg. Write a program to scan these texts for any extremely long sentences. What is the longest sentence you can find? What syntactic construction(s) are responsible for such long sentences?__

In [47]:
sents_gutenberg = [sent for file in gutenberg.fileids()[:5] for sent in gutenberg.sents(file)]

In [49]:
print(*sorted(sents_gutenberg, key = lambda x: len(x), reverse = True)[0])

15 : 21 And the uttermost cities of the tribe of the children of Judah toward the coast of Edom southward were Kabzeel , and Eder , and Jagur , 15 : 22 And Kinah , and Dimonah , and Adadah , 15 : 23 And Kedesh , and Hazor , and Ithnan , 15 : 24 Ziph , and Telem , and Bealoth , 15 : 25 And Hazor , Hadattah , and Kerioth , and Hezron , which is Hazor , 15 : 26 Amam , and Shema , and Moladah , 15 : 27 And Hazargaddah , and Heshmon , and Bethpalet , 15 : 28 And Hazarshual , and Beersheba , and Bizjothjah , 15 : 29 Baalah , and Iim , and Azem , 15 : 30 And Eltolad , and Chesil , and Hormah , 15 : 31 And Ziklag , and Madmannah , and Sansannah , 15 : 32 And Lebaoth , and Shilhim , and Ain , and Rimmon : all the cities are twenty and nine , with their villages : 15 : 33 And in the valley , Eshtaol , and Zoreah , and Ashnah , 15 : 34 And Zanoah , and Engannim , Tappuah , and Enam , 15 : 35 Jarmuth , and Adullam , Socoh , and Azekah , 15 : 36 And Sharaim , and Adithaim , and Gederah , and Gedero

In [50]:
# Well, the sent above seems to be from "bible-kvj.txt"
gutenberg.fileids()[:5]

['austen-emma.txt',
 'austen-persuasion.txt',
 'austen-sense.txt',
 'bible-kjv.txt',
 'blake-poems.txt']

__◑ Modify the functions init_wfst() and complete_wfst() so that the contents of each cell in the WFST is a set of non-terminal symbols rather than a single non-terminal.__

In [282]:
def init_wfst(tokens, grammar):
    numtokens = len(tokens)
    wfst = [[None for i in range(numtokens+1)] for j in range(numtokens+1)]
    for i in range(numtokens):
        productions = grammar.productions(rhs=tokens[i])
        wfst[i][i+1] = productions[0].lhs()
    return wfst

def complete_wfst(wfst, tokens, grammar, trace=False):
    index = dict((p.rhs(), p.lhs()) for p in grammar.productions())
    numtokens = len(tokens)
    for span in range(2, numtokens+1):
        for start in range(numtokens+1-span):
            end = start + span
            for mid in range(start+1, end):
                nt1, nt2 = wfst[start][mid], wfst[mid][end]
                if nt1 and nt2 and (nt1,nt2) in index:
                    wfst[start][end] = index[(nt1,nt2)]
                    if trace:
                        print("[%s] %3s [%s] %3s [%s] ==> [%s] %3s [%s]" % \
                        (start, nt1, mid, nt2, end, start, index[(nt1,nt2)], end))
    return wfst

def display(wfst, tokens):
    print('\nWFST ' + ' '.join(("%-4d" % i) for i in range(1, len(wfst))))
    for i in range(len(wfst)-1):
        print("%d   " % i, end=" ")
        for j in range(1, len(wfst)):
            print("%-4s" % (wfst[i][j] or '.'), end=" ")
        print()

__◑ Consider the algorithm in 4.4. Can you explain why parsing context-free grammar is proportional to n3, where n is the length of the input sentence.__

In [ ]:
# TODO

__◑ Process each tree of the Treebank corpus sample nltk.corpus.treebank and extract the productions with the help of Tree.productions(). Discard the productions that occur only once. Productions with the same left hand side, and similar right hand sides can be collapsed, resulting in an equivalent but more compact set of rules. Write code to output a compact grammar.__

In [119]:
def get_productions(treebank):
    """
    Given parsed sentences from a treebacnk corpus,
    the function returns non-terminal productions of the corpus. 
    If some productions have same LHS and their RHS is similar, then 
    they are merged. Otherwise, if no other similar RHS was found, that
    particular RHS is returned as one of possible non-terminals.
    
    The similarity of two RHS is determined by looking at the first and the last items of RHS.
    If they are the same, RHSes era said to be similar, and the longes of all similar RHSes
    is taken when they are merged. 
    
    """
    prod_dict = defaultdict(int)
    prods = set()
    prod_lhs_dict = defaultdict(lambda : defaultdict(int))
    prod_rhs_dict = defaultdict(lambda : defaultdict(set))
    lhs_dict = defaultdict(lambda : list())
    
    for tree in treebank:
        for prod in tree.productions():
            if prod.rhs()[0] != str:  # If it is a non-terminal, then get the production. Otehrwise discard
                prod_dict[prod] += 1
        
    for prod, freq in prod_dict.items():
        if freq > 1:
            prods.add(prod)
            
    for prod in prods:
        LHS = prod.lhs()
        if prod.rhs(): RHS = prod.rhs()
        RHS_0 = RHS[0]
        RHS_last = RHS[-1]
        prod_lhs_dict[LHS][(RHS_0, RHS_last)] += 1
        prod_rhs_dict[LHS][(RHS_0, RHS_last)].add(RHS)
    
    for lhs, dict_2 in prod_lhs_dict.items():
        for tuples, freq in dict_2.items():
            if freq > 1:
                lhs_dict[lhs].append(max(prod_rhs_dict[lhs][tuples]))
            elif freq == 1:
                lhs_dict[lhs] + list((prod_rhs_dict[lhs][tuples]))
            
    
    return lhs_dict

In [120]:
prods = get_productions(treebank.parsed_sents())

In [121]:
len(prods)  # 227 unique LHS values are found

227

In [122]:
list(prods.items())[0]

(VP,
 [(VBZ, SBAR, ,, S-ADV),
  (VBZ, PP-DIR),
  (VB, PRT, PP-CLR),
  (VBG, PP-CLR, PP-CLR),
  (VBN, S, PP),
  (VB, PP-DIR, PP-PRP),
  (VBD, PRT, PP-CLR),
  (VB, SBAR-ADV),
  (VBD, SBAR, PP-TMP),
  (VBD, SBAR),
  (VBN, PP-DIR, PP-DIR),
  (VBP, VP),
  (VBZ, S),
  (ADVP, VBP, NP),
  (VBG, PP),
  (VBZ, PP-MNR),
  (VBD, S),
  (VBG, PP-DIR),
  (VBN, PP-PRP),
  (VBP, NP-PRD, ,, S-ADV),
  (VBN, S, PP-TMP),
  (VBD, S-PRP),
  (VBP, RB, NP-PRD),
  (VBD, SBAR, ,, PP),
  (VB, ``, NP-PRD),
  (VBD, PP-CLR, ,, ADVP),
  (VBG, PP-LOC),
  (VB, NP, ADVP-MNR),
  (VBP, PP-CLR, PP-CLR),
  (VBN, S, ,, S-ADV),
  (VBN, S, SBAR-ADV),
  (VBZ, NP-PRD, ADVP-TMP),
  (VBZ, PP-CLR),
  (VB, S, PP-TMP),
  (VB, PP-CLR, PP),
  (VBG, S-ADV),
  (VB, ``, NP),
  (VB, SBAR-TMP),
  (VBP, PP),
  (VBD, VP),
  (VBD, ``, ADJP-PRD),
  (VBZ, ``, NP),
  (VB, PRT, NP, S-PRP),
  (VBN, PRT, NP),
  (VBD, SBAR, ,, SBAR-ADV),
  (VBG, PP-MNR),
  (VBG, NP, SBAR-TMP),
  (VBN, PP-LOC),
  (VBP, NP-PRD, ,, SBAR-ADV),
  (VP, CC, VP),
  (VBN, PRT,

__★ One common way of defining the subject of a sentence S in English is as the noun phrase that is the child of S and the sibling of VP. Write a function that takes the tree for a sentence and returns the subtree corresponding to the subject of the sentence. What should it do if the root node of the tree passed to this function is not S, or it lacks a subject?__

In [150]:
# If I got the question correctly, here is what is expected from me
def subject_finder(tree):
    for child in tree:
        if child.label().startswith("NP-SBJ"):
            print(child)

In [151]:
subject_finder(treebank.parsed_sents()[101])

(NP-SBJ
  (NP (DT A) (JJ full) (, ,) (JJ four-color) (NN page))
  (PP-LOC (IN in) (NP (NNP Newsweek))))


__★ Write a function that takes a grammar (such as the one defined in 3.1) and returns a random sentence generated by the grammar. (Use grammar.start() to find the start symbol of the grammar; grammar.productions(lhs) to get the list of productions from the grammar that have the specified left-hand side; and production.rhs() to get the right-hand side of a production.)__

In [152]:
grammar2 = nltk.CFG.fromstring("""
  S  -> NP VP
  NP -> Det Nom | PropN
  Nom -> Adj Nom | N
  VP -> VI Adj | VT NP | VS S | VT NP PP
  PP -> P NP
  PropN -> 'Buster' | 'Chatterer' | 'Joe' | 'Lee'
  Det -> 'the' | 'a'
  N -> 'bear' | 'squirrel' | 'tree' | 'fish' | 'log' | 'dog' | 'cat'
  Adj  -> 'angry' | 'frightened' |  'little' | 'tall'
  VI -> 'was'
  VT -> 'chased'  | 'saw'
  VS -> 'said' | 'thought' | 'put'
  P -> 'on'
  """)

In [212]:
def generate_from_grammar(grammar):
    
    start = grammar.start()
    prod_dict = defaultdict(lambda: list())
    
    for prod in grammar.productions():
        prod_dict[prod.lhs()].append(prod.rhs())
    
    next_rhs = prod_dict[start]
    
    def get_next(rhs):
        phrase = ""
        for tuples in rhs:
            if tuples[0] == nltk.grammar.Nonterminal:  # if RHS is non-terminal
                tuple_ = random.choice(rhs)
                if len(tuple_) == 1: 
                    get_next(prod_dict[tuple_[0]])
                else:
                    for i in tuple_:
                        get_next(prod_dict[i])
            elif tuples[0] == str:  # if RHS is not a non-terminal, add the word to "phrase"
                tuple_ = random.choice(rhs)
                phrase += tuple_[0]
        return phrase
    
        
    return get_next(next_rhs)

In [224]:
generate_from_grammar(grammar2)

''

__★ As we saw in 7., it is possible to collapse chunks down to their chunk label. When we do this for sentences involving the word gave, we find patterns such as the following:__

    gave NP
    gave up NP in NP
    gave NP up
    gave NP NP
    gave NP to NP
    
Use this method to study the complementation patterns of a verb of interest, and write suitable grammar productions. (This task is sometimes called lexical acquisition.)

Identify some English verbs that are near-synonyms, such as the dumped/filled/loaded example from earlier in this chapter. Use the chunking method to study the complementation patterns of these verbs. Create a grammar to cover these cases. Can the verbs be freely substituted for each other, or are their constraints? Discuss your findings.

In [228]:
conll_train_NP = conll2000.chunked_sents('train.txt')

In [247]:
print(conll_train_NP[3])

(S
  (NP This/DT)
  (VP has/VBZ increased/VBN)
  (NP the/DT risk/NN)
  (PP of/IN)
  (NP the/DT government/NN)
  (VP being/VBG forced/VBN to/TO increase/VB)
  (NP base/NN rates/NNS)
  (PP to/TO)
  (NP 16/CD %/NN)
  (PP from/IN)
  (NP their/PRP$ current/JJ 15/CD %/NN level/NN)
  (VP to/TO defend/VB)
  (NP the/DT pound/NN)
  ,/,
  (NP economists/NNS)
  and/CC
  (NP foreign/JJ exchange/NN market/NN analysts/NNS)
  (VP say/VBP)
  ./.)
